File: apps/api/src/lumi_api/constraint_validator/runtime.py
from __future__ import annotations

from typing import Any, Mapping

from .contracts import (
    ConstraintViolation,
    RuntimeConstraint,
    ValidationAdapters,
    ValidationMetrics,
    ValidationPhase,
    ValidationPolicy,
    ValidationReport,
)
from .impact import impact_set
from .projection import project_operation
from .validators import relevant_specs

_LOCKED_OPERATION_TYPES: dict[str, frozenset[str]] = {
    "LOCK_POSITION": frozenset({"MOVE_NODE"}),
    "LOCK_SIZE": frozenset({"RESIZE_NODE"}),
    "LOCK_ROTATION": frozenset({"ROTATE_NODE"}),
    "LOCK_TRANSFORM": frozenset({"MOVE_NODE", "RESIZE_NODE", "ROTATE_NODE"}),
    "LOCK_LAYER_ORDER": frozenset({"REORDER_NODE"}),
    "LOCK_PARENT": frozenset({"REPARENT_NODE"}),
    "LOCK_CONTENT": frozenset({"SET_TEXT", "REPLACE_ASSET", "DELETE_NODE"}),
    "LOCK_TEXT": frozenset({"SET_TEXT"}),
    "LOCK_ASSET": frozenset({"REPLACE_ASSET"}),
    "LOCK_STYLE": frozenset({"APPLY_STYLE", "SET_PROPERTY"}),
}


def _lock_applies(constraint: RuntimeConstraint, operation: Mapping[str, Any] | None) -> bool:
    if operation is None:
        return False
    allowed = _LOCKED_OPERATION_TYPES.get(constraint.type)
    return allowed is not None and operation.get("type") in allowed


def _health_score(
    constraints: tuple[RuntimeConstraint, ...], violations: tuple[ConstraintViolation, ...]
) -> float:
    weights = {"HARD": 5.0, "SOFT": 2.0, "ADVISORY": 1.0}
    active = [item for item in constraints if item.active]
    denominator = sum(weights[item.severity] for item in active)
    if denominator == 0:
        return 100.0
    failed_ids = {item.constraint_id for item in violations}
    penalty = sum(weights[item.severity] for item in active if item.constraint_id in failed_ids)
    return round(max(0.0, 100.0 * (1.0 - penalty / denominator)), 4)


def _sort_violations(values: list[ConstraintViolation]) -> tuple[ConstraintViolation, ...]:
    dedup = {item.violation_id: item for item in values}
    return tuple(
        sorted(
            dedup.values(),
            key=lambda item: (
                item.constraint_id,
                item.validator,
                item.affected_node_ids,
                item.violation_id,
            ),
        )
    )


def validate_constraints(
    document: Mapping[str, Any],
    constraints: tuple[RuntimeConstraint, ...],
    *,
    operation: Mapping[str, Any] | None = None,
    phase: ValidationPhase = "preflight",
    adapters: ValidationAdapters | None = None,
    policy: ValidationPolicy | None = None,
    force_full: bool = False,
) -> ValidationReport:
    active_policy = policy or ValidationPolicy()
    active_adapters = adapters or ValidationAdapters()
    candidate = project_operation(document, operation) if operation is not None else dict(document)
    impact, fallback = impact_set(
        candidate,
        operation,
        constraints,
        active_policy,
        force_full=force_full or phase == "export",
    )
    violations: list[ConstraintViolation] = []
    validators_run: set[str] = set()
    for constraint in sorted(
        (item for item in constraints if item.active), key=lambda item: item.constraint_id
    ):
        for spec in relevant_specs(constraint):
            if spec.name == "LockedRegionValidator" and not _lock_applies(
                constraint, operation
            ):
                continue
            if spec.name == "ExportDimensionValidator" and phase != "export":
                continue
            validators_run.add(spec.name)
            violations.extend(
                spec.fn(candidate, constraint, impact, active_adapters, active_policy)
            )
    ordered = _sort_violations(violations)
    blocking = sum(1 for item in ordered if item.blocking)
    unavailable = any(item.unavailable for item in ordered)
    if blocking:
        status = "BLOCKED"
    elif unavailable:
        status = "VALIDATION_UNAVAILABLE"
    elif ordered:
        status = "WARN"
    else:
        status = "PASS"
    return ValidationReport(
        status=status,
        violations=ordered,
        hard_pass=blocking == 0,
        health_score=_health_score(constraints, ordered),
        metrics=ValidationMetrics(
            validators_run=tuple(sorted(validators_run)),
            nodes_scanned=len(impact),
            violations_count=len(ordered),
            blocking_count=blocking,
            fallback_full_scan=fallback,
        ),
    )


def validate_batch(
    document: Mapping[str, Any],
    constraints: tuple[RuntimeConstraint, ...],
    operations: tuple[Mapping[str, Any], ...],
    *,
    adapters: ValidationAdapters | None = None,
    policy: ValidationPolicy | None = None,
) -> ValidationReport:
    working: Mapping[str, Any] = document
    all_violations: list[ConstraintViolation] = []
    validators: set[str] = set()
    nodes_scanned = 0
    fallback = False
    for operation in operations:
        report = validate_constraints(
            working,
            constraints,
            operation=operation,
            adapters=adapters,
            policy=policy,
        )
        all_violations.extend(report.violations)
        validators.update(report.metrics.validators_run)
        nodes_scanned += report.metrics.nodes_scanned
        fallback = fallback or report.metrics.fallback_full_scan
        working = project_operation(working, operation)
    ordered = _sort_violations(all_violations)
    blocking = sum(1 for item in ordered if item.blocking)
    unavailable = any(item.unavailable for item in ordered)
    if blocking:
        status = "BLOCKED"
    elif unavailable:
        status = "VALIDATION_UNAVAILABLE"
    elif ordered:
        status = "WARN"
    else:
        status = "PASS"
    return ValidationReport(
        status=status,
        violations=ordered,
        hard_pass=blocking == 0,
        health_score=_health_score(constraints, ordered),
        metrics=ValidationMetrics(
            validators_run=tuple(sorted(validators)),
            nodes_scanned=nodes_scanned,
            violations_count=len(ordered),
            blocking_count=blocking,
            fallback_full_scan=fallback,
        ),
    )


def validate_export(
    document: Mapping[str, Any],
    constraints: tuple[RuntimeConstraint, ...],
    *,
    adapters: ValidationAdapters | None = None,
    policy: ValidationPolicy | None = None,
) -> ValidationReport:
    return validate_constraints(
        document,
        constraints,
        phase="export",
        adapters=adapters,
        policy=policy,
        force_full=True,
    )


def to_ir_preflight_issues(report: ValidationReport) -> tuple[dict[str, Any], ...]:
    return tuple(
        {
            "code": "IR_CONSTRAINT_FAILED",
            "message": item.message,
            "node_ids": list(item.affected_node_ids),
        }
        for item in report.violations
        if item.blocking
    )

File: packages/design-constraints/src/validator/runtime.ts
import { impactSet } from "./impact";
import { projectOperation } from "./projection";
import type {
  ConstraintViolation,
  DesignDocumentLike,
  DesignOperationLike,
  IrIssueLike,
  RuntimeConstraint,
  ValidationAdapters,
  ValidationPhase,
  ValidationPolicy,
  ValidationReport,
} from "./types";
import { VALIDATOR_SPECS } from "./validators";

const DEFAULT_POLICY: ValidationPolicy = {
  incremental_full_scan_ratio: 0.4,
  incremental_full_scan_node_limit: 500,
  unavailable_hard_blocks: true,
  max_auto_fix_rounds: 1,
};

const LOCKED_OPERATION_TYPES: Readonly<Record<string, readonly string[]>> = {
  LOCK_POSITION: ["MOVE_NODE"],
  LOCK_SIZE: ["RESIZE_NODE"],
  LOCK_ROTATION: ["ROTATE_NODE"],
  LOCK_TRANSFORM: ["MOVE_NODE", "RESIZE_NODE", "ROTATE_NODE"],
  LOCK_LAYER_ORDER: ["REORDER_NODE"],
  LOCK_PARENT: ["REPARENT_NODE"],
  LOCK_CONTENT: ["SET_TEXT", "REPLACE_ASSET", "DELETE_NODE"],
  LOCK_TEXT: ["SET_TEXT"],
  LOCK_ASSET: ["REPLACE_ASSET"],
  LOCK_STYLE: ["APPLY_STYLE", "SET_PROPERTY"],
};

function lockApplies(
  constraint: RuntimeConstraint,
  operation: DesignOperationLike | undefined,
): boolean {
  if (!operation) return false;
  return LOCKED_OPERATION_TYPES[constraint.type]?.includes(operation.type) ?? false;
}

function healthScore(
  constraints: readonly RuntimeConstraint[],
  violations: readonly ConstraintViolation[],
): number {
  const weights = { HARD: 5, SOFT: 2, ADVISORY: 1 } as const;
  const active = constraints.filter((item) => item.active !== false);
  const denominator = active.reduce((sum, item) => sum + weights[item.severity], 0);
  if (!denominator) return 100;
  const failed = new Set(violations.map((item) => item.constraint_id));
  const penalty = active.reduce(
    (sum, item) => sum + (failed.has(item.constraint_id) ? weights[item.severity] : 0),
    0,
  );
  return Math.round(Math.max(0, 100 * (1 - penalty / denominator)) * 10_000) / 10_000;
}

function sortViolations(values: readonly ConstraintViolation[]): readonly ConstraintViolation[] {
  const dedup = new Map(values.map((item) => [item.violation_id, item]));
  return [...dedup.values()].sort((a, b) => {
    const left = `${a.constraint_id}\u0000${a.validator}\u0000${a.affected_node_ids.join(",")}\u0000${a.violation_id}`;
    const right = `${b.constraint_id}\u0000${b.validator}\u0000${b.affected_node_ids.join(",")}\u0000${b.violation_id}`;
    return left < right ? -1 : left > right ? 1 : 0;
  });
}

export function validateConstraints(
  document: DesignDocumentLike,
  constraints: readonly RuntimeConstraint[],
  options: {
    readonly operation?: DesignOperationLike;
    readonly phase?: ValidationPhase;
    readonly adapters?: ValidationAdapters;
    readonly policy?: ValidationPolicy;
    readonly forceFull?: boolean;
  } = {},
): ValidationReport {
  const phase = options.phase ?? "preflight";
  const policy = { ...DEFAULT_POLICY, ...(options.policy ?? {}) };
  if ((policy.max_auto_fix_rounds ?? 1) !== 1) {
    throw new Error("CONSTRAINT_VALIDATOR_REPAIR_LOOP_FORBIDDEN");
  }
  const adapters = options.adapters ?? {};
  const candidate = options.operation ? projectOperation(document, options.operation) : document;
  const impact = impactSet(
    candidate,
    options.operation,
    constraints,
    policy,
    options.forceFull === true || phase === "export",
  );
  const violations: ConstraintViolation[] = [];
  const validators = new Set<string>();
  for (const constraint of [...constraints]
    .filter((item) => item.active !== false)
    .sort((a, b) => a.constraint_id.localeCompare(b.constraint_id))) {
    for (const spec of VALIDATOR_SPECS) {
      if (!spec.constraintTypes.includes(constraint.type)) continue;
      if (spec.name === "LockedRegionValidator" && !lockApplies(constraint, options.operation)) {
        continue;
      }
      if (spec.name === "ExportDimensionValidator" && phase !== "export") continue;
      validators.add(spec.name);
      violations.push(...spec.fn(candidate, constraint, impact.nodeIds, adapters, policy));
    }
  }
  const ordered = sortViolations(violations);
  const blocking = ordered.filter((item) => item.blocking).length;
  const unavailable = ordered.some((item) => item.unavailable);
  const status = blocking
    ? "BLOCKED"
    : unavailable
      ? "VALIDATION_UNAVAILABLE"
      : ordered.length
        ? "WARN"
        : "PASS";
  return {
    status,
    violations: ordered,
    hard_pass: blocking === 0,
    health_score: healthScore(constraints, ordered),
    metrics: {
      validators_run: [...validators].sort(),
      nodes_scanned: impact.nodeIds.size,
      violations_count: ordered.length,
      blocking_count: blocking,
      fallback_full_scan: impact.fallbackFullScan,
    },
  };
}

export function validateBatch(
  document: DesignDocumentLike,
  constraints: readonly RuntimeConstraint[],
  operations: readonly DesignOperationLike[],
  options: {
    readonly adapters?: ValidationAdapters;
    readonly policy?: ValidationPolicy;
  } = {},
): ValidationReport {
  let working = document;
  const violations: ConstraintViolation[] = [];
  const validators = new Set<string>();
  let nodesScanned = 0;
  let fallback = false;
  for (const operation of operations) {
    const report = validateConstraints(working, constraints, {
      operation,
      ...(options.adapters ? { adapters: options.adapters } : {}),
      ...(options.policy ? { policy: options.policy } : {}),
    });
    violations.push(...report.violations);
    for (const validator of report.metrics.validators_run) validators.add(validator);
    nodesScanned += report.metrics.nodes_scanned;
    fallback ||= report.metrics.fallback_full_scan;
    working = projectOperation(working, operation);
  }
  const ordered = sortViolations(violations);
  const blocking = ordered.filter((item) => item.blocking).length;
  const unavailable = ordered.some((item) => item.unavailable);
  return {
    status: blocking
      ? "BLOCKED"
      : unavailable
        ? "VALIDATION_UNAVAILABLE"
        : ordered.length
          ? "WARN"
          : "PASS",
    violations: ordered,
    hard_pass: blocking === 0,
    health_score: healthScore(constraints, ordered),
    metrics: {
      validators_run: [...validators].sort(),
      nodes_scanned: nodesScanned,
      violations_count: ordered.length,
      blocking_count: blocking,
      fallback_full_scan: fallback,
    },
  };
}

export function validateExport(
  document: DesignDocumentLike,
  constraints: readonly RuntimeConstraint[],
  options: { readonly adapters?: ValidationAdapters; readonly policy?: ValidationPolicy } = {},
): ValidationReport {
  return validateConstraints(document, constraints, {
    phase: "export",
    forceFull: true,
    ...(options.adapters ? { adapters: options.adapters } : {}),
    ...(options.policy ? { policy: options.policy } : {}),
  });
}

export function reportToIrIssues(
  report: ValidationReport,
  operationId?: string,
): readonly IrIssueLike[] {
  return report.violations
    .filter((item) => item.blocking)
    .map((item) => ({
      code: "IR_CONSTRAINT_FAILED" as const,
      message: item.message,
      node_ids: item.affected_node_ids,
      ...(operationId ? { operation_id: operationId } : {}),
    }));
}

export function createIrPreflight(
  constraints: readonly RuntimeConstraint[],
  options: { readonly adapters?: ValidationAdapters; readonly policy?: ValidationPolicy } = {},
): (document: DesignDocumentLike, operation: DesignOperationLike) => readonly IrIssueLike[] {
  return (document, operation) =>
    reportToIrIssues(
      validateConstraints(document, constraints, {
        operation,
        ...(options.adapters ? { adapters: options.adapters } : {}),
        ...(options.policy ? { policy: options.policy } : {}),
      }),
      operation.operation_id,
    );
}
